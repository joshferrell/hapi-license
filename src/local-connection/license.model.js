import Sequelize from 'sequelize';

const createLicenseModel = connection => connection.define('license', {
    id: {
        type: Sequelize.UUID,
        primaryKey: true,
        defaultValue: Sequelize.UUIDV4()
    },
    email: {
        type: Sequelize.STRING,
        allowNull: false,
        validate: {
            notEmpty: true,
            isEmail: true
        }
    },
    productId: {
        type: Sequelize.STRING,
        allowNull: false,
        validate: {
            notEmpty: true
        }
    },
    computerId: {
        type: Sequelize.STRING,
        allowNull: false,
        validate: {
            notEmpty: true
        }
    },
    computerUsername: {
        type: Sequelize.STRING,
        allowNull: false,
        validate: {
            notEmpty: true
        }
    },
    computerOS: {
        type: Sequelize.STRING,
        allowNull: false,
        validate: {
            isIn: [['Windows', 'MacOS', 'Linux']]
        }
    },
    computerName: {
        type: Sequelize.STRING,
        allowNull: false,
        validate: {
            notEmpty: true
        }
    },
    expiresAt: {
        type: Sequelize.DATE,
        allowNull: false
    },
    license: {
        type: Sequelize.TEXT,
        allowNull: false
    }
});

export default createLicenseModel;
