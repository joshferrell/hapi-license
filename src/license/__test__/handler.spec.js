import Sequelize from 'sequelize';
import boom from 'boom';
import { dissoc, pipe } from 'ramda';
import { createLicenseModel } from '../../local-connection';
import {
    createFetchLicenses,
    createDeleteLicense,
    createNewLicense,
    createFetchLicense,
    createUpdateLicense
} from '../license.handler';

describe('license handlers', () => {
    let LicenseModel;
    let request;
    let reply;

    const LicenseModelBad = {
        create: () => Promise.reject(new Error('bad to the bone')),
        findAll: () => Promise.reject(new Error('bad to the bone')),
        destroy: () => Promise.reject(new Error('bad to the bone')),
        find: () => Promise.reject(new Error('bad to the bone'))
    };

    const simpleLicense = {
        id: 'c797c6bc-ef64-4203-9956-a3a90c0c0927',
        email: 'example@gmail.com',
        productId: 'simple-product',
        computerId: 'unique-machine-name',
        computerUsername: 'paul',
        computerOS: 'Windows',
        computerName: 'paul\'s laptop',
        expiresAt: '2099-01-01',
        license: 'hi I\'m paul',
        createdAt: '2017-01-01'
    };

    beforeAll(() => {
        process.env.SERVER_URL = 'localhost:3000';
        process.env.SERVER_SCHEME = 'http';
        const connection = new Sequelize('database', 'username', 'password', {
            dialect: 'sqlite'
        });

        LicenseModel = createLicenseModel(connection);
    });

    beforeEach(async () => {
        await LicenseModel.sync({ force: true });
        request = {
            log: jest.fn(),
            auth: {
                credentials: {
                    email: 'example@gmail.com',
                    app_metadata: {
                        licenseCount: 1,
                        licenseTotal: 5
                    }
                }
            }
        };
        reply = jest.fn();
    });

    describe('fetching licenses', () => {
        beforeEach(() => LicenseModel.create(simpleLicense));

        it('should return an array of licenses with an id url', async () => {
            const fetchLicenses = createFetchLicenses(LicenseModel);
            await fetchLicenses(request, reply);
            expect(reply).toHaveBeenCalled();
            expect(reply).toHaveBeenCalledTimes(1);

            const licenses = reply.mock.calls[0][0]
                .licenses
                .map(pipe(
                    dissoc('updatedAt'),
                    dissoc('createdAt')
                ));
            const { meta } = reply.mock.calls[0][0];

            expect(licenses).toHaveLength(1);
            expect(licenses).toMatchSnapshot();
            expect(meta.licenseCount).toBe(1);
            expect(meta.licenseTotal).toBe(5);
        });

        it('should log and reply error if fatal error', async () => {
            const fetchLicenses = createFetchLicenses(LicenseModelBad);
            await fetchLicenses(request, reply);
            const errorLog = request.log.mock.calls[0];

            expect(reply).toHaveBeenCalled();
            expect(reply).toHaveBeenCalledTimes(1);
            expect(reply).toHaveBeenCalledWith(boom.badImplementation());

            expect(request.log).toHaveBeenCalled();
            expect(request.log).toHaveBeenCalledTimes(1);
            expect(errorLog[0]).toEqual('error');
            expect(errorLog[1]).toMatchSnapshot();
        });
    });

    describe.skip('creating license', () => {

    });

    describe('delete license', () => {
        beforeEach(() => {
            const { id } = simpleLicense;
            request.params = { id };
            return LicenseModel.create(simpleLicense);
        });

        it('should delete an existing license', async () => {
            const deleteLicense = createDeleteLicense(LicenseModel);
            await deleteLicense(request, reply);

            expect(reply).toHaveBeenCalled();
            expect(reply).toHaveBeenCalledTimes(1);
            expect(reply).toHaveBeenCalledWith({
                success: true
            });
        });

        it('should not remove a license if there is no access', async () => {
            const deleteLicense = createDeleteLicense(LicenseModel);
            request.auth.credentials.email = 'bad@superbad.io';

            await deleteLicense(request, reply);

            expect(reply).toHaveBeenCalled();
            expect(reply).toHaveBeenCalledTimes(1);
            expect(reply).toHaveBeenCalledWith(boom.badRequest('license does not exist'));
        });

        it('should log and reply 500 if error', async () => {
            const deleteLicense = createDeleteLicense(LicenseModelBad);
            await deleteLicense(request, reply);
            const errorLog = request.log.mock.calls[0][1];

            expect(reply).toHaveBeenCalled();
            expect(reply).toHaveBeenCalledTimes(1);
            expect(reply).toHaveBeenCalledWith(boom.badImplementation());

            expect(request.log).toHaveBeenCalled();
            expect(request.log).toHaveBeenCalledTimes(1);
            expect(request.log.mock.calls[0][0]).toEqual('error');
            expect(errorLog).toMatchSnapshot();
        });
    });

    describe.skip('fetch single license', () => {
        
    });

    describe.skip('update licenses', () => {

    });
});
