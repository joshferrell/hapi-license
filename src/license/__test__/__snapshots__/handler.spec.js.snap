// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`license handlers delete license should log and reply 500 if error 1`] = `
Object {
  "err": [Error: bad to the bone],
  "msg": "unable to delete license",
}
`;

exports[`license handlers fetch single license should be able to fetch a single license 1`] = `
Object {
  "computerId": "unique-machine-name",
  "computerName": "paul's laptop",
  "computerOS": "Windows",
  "computerUsername": "paul",
  "email": "example@gmail.com",
  "expiresAt": "2099-01-01 00:00:00.000 +00:00",
  "id": "c797c6bc-ef64-4203-9956-a3a90c0c0927",
  "license": "hi I'm paul",
  "productId": "simple-product",
  "url": "http://localhost:3000/license/c797c6bc-ef64-4203-9956-a3a90c0c0927",
}
`;

exports[`license handlers fetch single license should be able to fetch a single license 2`] = `
Object {
  "licenseCount": 1,
  "licenseTotal": 5,
}
`;

exports[`license handlers fetch single license should return 500 and log an error if critical error ocurrs 1`] = `
Object {
  "err": [Error: bad to the bone],
  "msg": "fetch inidividual license",
}
`;

exports[`license handlers fetching licenses should log and reply error if fatal error 1`] = `
Object {
  "err": [Error: bad to the bone],
  "msg": "unable to fetch licenses",
}
`;

exports[`license handlers fetching licenses should return an array of licenses with an id url 1`] = `
Array [
  Object {
    "computerId": "unique-machine-name",
    "computerName": "paul's laptop",
    "computerOS": "Windows",
    "computerUsername": "paul",
    "email": "example@gmail.com",
    "expiresAt": "2099-01-01 00:00:00.000 +00:00",
    "id": "c797c6bc-ef64-4203-9956-a3a90c0c0927",
    "license": "hi I'm paul",
    "productId": "simple-product",
    "url": "http://localhost:3000/license/c797c6bc-ef64-4203-9956-a3a90c0c0927",
  },
]
`;
